cmd/otel-allocator/allocation/allocator.go

```diff
@@ -88,119 +88,129 @@ func (allocator *Allocator) Collectors() map[string]*collector {
 	for _, i := range collectors {
 		allocator.collectors[i] = &collector{Name: i, NumTargets: 0}
 	}
-	collectorsAllocatable.Set(float64(len(collectors)))
-}
-
-// AllocateTargets removes outdated targets and adds new ones from
-// waitingTargets. This method needs to be called to process the new target
-// updates. Until it is called, old targets will be served.
-func (allocator *Allocator) AllocateTargets() {
-	allocator.m.Lock()
-	timer := prometheus.NewTimer(timeToAssign.WithLabelValues("AllocateTargets"))
-	defer timer.ObserveDuration()
-	defer allocator.m.Unlock()
-	allocator.removeOutdatedTargets()
-	allocator.processWaitingTargets()
-}
 
-// ReallocateCollectors reallocates the targets among the new collector instances.
-func (allocator *Allocator) ReallocateCollectors() {
-	allocator.m.Lock()
-	timer := prometheus.NewTimer(timeToAssign.WithLabelValues("ReallocateCollectors"))
-	defer timer.ObserveDuration()
-	defer allocator.m.Unlock()
-	allocator.targetItems = make(map[string]*TargetItem)
-	allocator.processWaitingTargets()
-}
-
-// removeOutdatedTargets removes targets that are no longer available. This
-// method is called after a lock has been acquired in ReallocateCollectors().
-func (allocator *Allocator) removeOutdatedTargets() {
-	for k := range allocator.targetItems {
-		if _, ok := allocator.targetsWaiting[k]; !ok {
-			allocator.collectors[allocator.targetItems[k].Collector.Name].NumTargets--
-			delete(allocator.targetItems, k)
-		}
-	}
-}
-
-// processWaitingTargets processes the newly set targets. This method is called
-// after a lock has been acquired in AllocateTargets() or ReallocateCollectors().
-func (allocator *Allocator) processWaitingTargets() {
-	for k, v := range allocator.targetsWaiting {
-		if _, ok := allocator.targetItems[k]; !ok {
-			col := allocator.findNextCollector()
-			allocator.targetItems[k] = &v
-			targetItem := TargetItem{
-				JobName:   v.JobName,
-				Link:      LinkJSON{fmt.Sprintf("/jobs/%s/targets", url.QueryEscape(v.JobName))},
-				TargetURL: v.TargetURL,
-				Label:     v.Label,
-				Collector: col,
-			}
-			col.NumTargets++
-			targetsPerCollector.WithLabelValues(col.Name).Set(float64(col.NumTargets))
-			allocator.targetItems[v.hash()] = &targetItem
-		}
+	// Re-Allocate the existing targets
+	for _, item := range allocator.targetItems {
+		allocator.assignTargetToNextCollector(item)
 	}
-}
 
-// findNextCollector finds the next collector with fewer number of targets.
-// This method is called from within processWaitingTargets(), whose caller
-// acquires the needed lock.
-func (allocator *Allocator) findNextCollector() *collector {
-	var col *collector
-	for _, v := range allocator.collectors {
-		// If the initial collector is empty, set the initial collector to the first element of map
-		if col == nil {
-			col = v
-		} else {
-			if v.NumTargets < col.NumTargets {
-				col = v
-			}
-		}
-
-	}
-	return col
+	collectorsAllocatable.Set(float64(len(collectors)))
 }
 
```

Comment:
```suggestion

// getCollectorChanges returns the missing and surplus collectors respectively.
// This method is called from within SetCollectors, which acquires the needed lock.
func (allocator *Allocator) getCollectorChanges(collectors []string) ([]string, []string) {
	var missing []string
	var surplus []string
	for _, s := range collectors {
		if _, found := allocator.collectors[s]; !found {
			missing = append(missing, s)
		}
	}
	for k := range allocator.collectors {
		found := false
		for _, s := range collectors {
			if k == s {
				found = true
			}
		}
		if !found {
			surplus = append(surplus, k)
		}
	}
	return missing, surplus
}

```

Reply:
Thank you for the suggestion! I implemented this method, and it allows me to replace the other method i made "allCollectorsPresent" because this essentially checks the same thing.